src/AppInstallerCLICore/Workflows/PinFlow.cpp

```diff
@@ -0,0 +1,247 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+#include "pch.h"
+#include "Resources.h"
+#include "PinFlow.h"
+#include "TableOutput.h"
+#include "Microsoft/PinningIndex.h"
+#include "Microsoft/SQLiteStorageBase.h"
+#include "winget/RepositorySearch.h"
+
+using namespace AppInstaller::Repository;
+using namespace AppInstaller::Repository::Microsoft;
+
+namespace AppInstaller::CLI::Workflow
+{
+    namespace
+    {
+        // Creates a Pin appropriate for the context based on the arguments provided
+        Pinning::Pin CreatePin(Execution::Context& context, std::string_view packageId, std::string_view sourceId, const std::string& installedVersion)
+        {
+            if (context.Args.Contains(Execution::Args::Type::GatedVersion))
+            {
+                return Pinning::Pin::CreateGatingPin({ packageId, sourceId }, context.Args.GetArg(Execution::Args::Type::GatedVersion));
+            }
+            else if (context.Args.Contains(Execution::Args::Type::BlockingPin))
+            {
+                return Pinning::Pin::CreateBlockingPin({ packageId, sourceId }, { installedVersion });
+            }
+            else
+            {
+                return Pinning::Pin::CreatePinningPin({ packageId, sourceId }, { installedVersion });
+            }
+        }
+    }
+
+    void OpenPinningIndex(Execution::Context& context)
+    {
+        auto indexPath = Runtime::GetPathTo(Runtime::PathName::LocalState) / "pinning.db";
+
+        AICLI_LOG(CLI, Info, << "Opening pinning index");
+
+        auto pinningIndex = std::filesystem::exists(indexPath) ?

```

Comment:
We have issues before with settings where a folder with same name exists and code will crash when opening them. We should have better handling for corner cases here.

Reply:
Do you have a pointer to how we handle that for other files?

Reply:
https://github.com/microsoft/winget-cli/pull/2767 is where Ruben fixes issues when settings.json is a directory, but mainly we try to catch exceptions, remove bad one and recrate an empty new one.

Reply:
I changed it to have a softer error if we cannot open the database instead of crashing. I'll also update the follow-up PR so that if we cannot open it, we just don't consider pins during upgrade/other commands.

I don't think this will cause the situation of making a directory in the location of the file because I'm not getting the full path with Runtime::GetPathTo, rather, I'm getting the path to LocalState and adding to it.